build_zng.rs

```diff
@@ -1,58 +1,691 @@
+        if let Some((begin, end)) = line.split_once("@ZLIB_SYMBOL_PREFIX@") {
+            write(&[
+                IoSlice::new(begin.as_bytes()),
+                IoSlice::new(end.as_bytes()),
+                IoSlice::new(b"\n"),
+            ]);
+        } else {
+            write(&[IoSlice::new(line.as_bytes()), IoSlice::new(b"\n")]);
+        }
+
+        on_line(line);
+    }
+}
+
+#[derive(Default)]
+struct AppendState {
+    flags: BTreeSet<&'static str>,
+    defines: BTreeSet<&'static str>,
+    files: BTreeSet<&'static str>,
+}
+
+#[derive(Debug, Hash, Copy, Clone)]
+struct TargetFeature {
+    check: &'static str,
+    msvc_flags: &'static [&'static str],
+    flags: &'static [&'static str],
+    defines: &'static [&'static str],
+    files: &'static [&'static str],
+}
+
+struct Ctx<'ctx> {
+    cfg: &'ctx mut cc::Build,
+    root: Option<&'ctx str>,
+    enabled: Option<&'ctx BTreeSet<&'ctx str>>,
+    msvc: bool,
+    append: AppendState,
+    cache: Option<BTreeMap<u64, bool>>,
+}
+
+impl<'ctx> Ctx<'ctx> {
+    fn enable(&mut self, tf: TargetFeature) -> bool {
+        if let Some(enabled) = self.enabled {
+            if !enabled.contains(tf.check) {
+                return false;
+            }
+        }
+
+        self.push(tf);
+        true
+    }
+
+    fn compile_check(&mut self, subdir: &str, tf: TargetFeature) -> bool {
+        let dst = Self::compile_check_root();
+        fs::create_dir_all(&dst).unwrap();
+
+        if let Some(res) = self.check_cache(&tf) {
+            println!("cargo:warning=using cached compile check");
+            if res {
+                self.push(tf);
+            }
+
+            return res;
+        }
+
+        let mut cmd = self.cfg.get_compiler().to_command();
+
+        // compile only
+        cmd.arg(if self.msvc { "/c" } else { "-c" });
+
+        // set output file so we don't pollute the cwd
+        {
+            let path = {
+                let mut pb = dst.join(tf.check);
+                pb.set_extension(if self.msvc { "obj" } else { "o" });
+                pb
+            };
+            if self.msvc {
+                cmd.arg(format!("/Fo\"{}\"", path.display()));
+            } else {
+                cmd.arg("-o");
+                cmd.arg(path);
+            }
+        }
+
+        let flags = if self.msvc { tf.msvc_flags } else { tf.flags };
+        cmd.args(flags);
+
+        let path = {
+            let mut p = "src/compile_check/".to_owned();
+            if !subdir.is_empty() {
+                p.push_str(subdir);
+                p.push('/');
+            }
+
+            p.push_str(tf.check);
+            p.push_str(".c");
+            p
+        };
+
+        cmd.arg(path);
+
+        let output = cmd.output().expect("failed to run compiler");
+
+        let result = if !output.status.success() {
+            println!("cargo:warning=failed to compile check '{tf:?}'");
+
+            for line in String::from_utf8(output.stderr).unwrap().lines() {
```

Comment:
I think panicking on non utf-8 character isn't very robust:

```suggestion
            for line in String::from_utf8_lossy(output.stderr).lines() {
```

Reply:
There is no case where a compiler would output non-utf8 strings, but sure.